### PR TITLE
added hat

### PR DESCRIPTION
**Pull request by qadan**

### CODE REVIEW

File: modules/derivatives_for_hats/derivatives_for_hats.module [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Attempts to generate derivatives from the object OBJ datastream.
+ *
+ * @param AbstractObject $object
+ *   The object to derive for.
+ */
+function derivatives_for_hats_generate_derivatives(AbstractObject $object) {
+  if ($object['OBJ']) {
+    // Recreates islandora_do_batch_derivatives() in immediate mode.
```
Comment (willtp87) [inline modules/derivatives_for_hats/derivatives_for_hats.module:34]: This is great you should say why in a README, cause if I saw this and didn't know I  would just ctr-x

File: modules/hats_in_a_book/hats_in_a_book.module [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * @param string $book
+ *   The PID of the book to add the pages to.
+ */
+function hats_in_a_book_add_to_book(AbstractObject $object, $book) {
+  if (in_array('islandora:pageCModel', $object->models)) {
+    // Add the page to the book.
+    module_load_include('inc', 'islandora_paged_content', 'includes/utilities');
+    $current_page = cache_get('hats_in_a_book_current_page');
+    $current_page = $current_page->data;
+    islandora_paged_content_set_relationship($object->relationships, ISLANDORA_RELS_EXT_URI, 'isPageOf', $book);
```
Comment (willtp87) [inline modules/hats_in_a_book/hats_in_a_book.module]: Should call https://github.com/Islandora/islandora_paged_content/blob/7.x/includes/utilities.inc#L1127

File: modules/log_hat_downloads/log_hat_downloads.info [hunk 3]
```diff
@@ -0,0 +1,9 @@
+name = Log Hat Downloads
+description = Pairs with islandora_usage_stats to log a download for each ingested OBJ datastream.
+core = 7.x
+version = 7.x-dev
+dependencies[] = ten_million_with_a_hat
+dependencies[] = islandora_usage_stats
+dependencies[] = objs_for_hats
+package = Ten Million with a Hat
+
```
Comment (willtp87) [inline modules/log_hat_downloads/log_hat_downloads.info]: Extra new line.

**Comment by willtp87** [comment]
Dan says usage stats stuff doesn't work and needs a refactor.

**Comment by willtp87** [comment]
Readmes should be updated to current format and add in dgi info.

File: modules/log_hat_views/log_hat_views.info [hunk 4]
```diff
@@ -0,0 +1,8 @@
+name = Log Hat Views
+description = Pairs with islandora_usage_stats to log a view for each ingested object.
+core = 7.x
+version = 7.x-dev
+dependencies[] = ten_million_with_a_hat
+dependencies[] = islandora_usage_stats
+package = Ten Million with a Hat
+
```
Comment (willtp87) [inline modules/log_hat_views/log_hat_views.info]: Watch your new lines.

File: modules/mods_for_hats/mods_for_hats.module [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+/**
+ * Adds a randomized MODS string to a hat.
+ *
+ * @param AbstractObject $object
+ *   The object to put the MODS on.
+ */
+function mods_for_hats_put_mods_on_hat(AbstractObject $object) {
+  $mods = $object->constructDatastream('MODS', 'M');
+  $mods->mimetype = 'text/xml';
```
Comment (willtp87) [inline modules/mods_for_hats/mods_for_hats.module]: Wrong mime.

File: modules/tiffs_for_page_hats/tiffs_for_page_hats.info [hunk 6]
```diff
@@ -0,0 +1,7 @@
+name = Tiffs for Page Hats
+description = Pairs with the paged content solution pack , as well as ImageMagick and fortune, to generate tiffs for page objects.
```
Comment (willtp87) [inline modules/tiffs_for_page_hats/tiffs_for_page_hats.info:2]: Needs cowsay.

File: ten_million_with_a_hat.api.php [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * @param AbstractObject $object
+ *   The object being worked with.
+ * @param string $new_colour
+ *   The new colour to use.
+ *
+ * @return string
+ *   The message to display.
+ */
+function hat_colour_change_colour(AbstractObject $object, $new_colour) {
```
Comment (willtp87) [inline ten_million_with_a_hat.api.php]: Should look at drupal docs for how callbacks should be documented.

File: ten_million_with_a_hat.module [hunk 8]
```diff
@@ -2,5 +2,298 @@
 
 /**
  * @file
- * To be a module.
+ * 🔟Ⓜ️ w/ 🎩.
```
Comment (willtp87) [inline ten_million_with_a_hat.module]: Lets not do this some systems will barf.

File: ten_million_with_a_hat.info [hunk 9]
```diff
@@ -3,4 +3,6 @@ description = Ingests lots of objects.
 core = 7.x
 version = 7.x-dev
 dependencies[] = islandora
-package = Islandora Tools
+dependencies[] = islandora_basic_collection
+dependencies[] = simpletest
```
Comment (willtp87) [inline ten_million_with_a_hat.info]: Kill it!

**Comment by willtp87** [comment]
Dan wants to pass the batch context through the hooks.  That's rather unique so I'd make the documentation rather complete.

**Comment by willtp87** [comment]
READMEs should look more like https://gist.github.com/willtp87/8bdc67e0d470a07d1eea#file-readme-md for non-foundation work.

**Comment by qadan** [comment]
Note that while I recognize that the sandbox stuff COULD be used for objs_for_hats, the current implementation there is working fine and it would be pretty messy to get the sandbox in there, so I'm letting it be.

**Comment by qadan** [comment]
sheezus christmas

**Pull request merged**